sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDbStateStoreProvider.scala

```diff
@@ -0,0 +1,652 @@
+        if (earliestVersionToRetain > 0) {
+          val earliestFileToRetain = filesForVersion(files, earliestVersionToRetain).head
+          val filesToDelete = files.filter(_.version < earliestFileToRetain.version)
+          val (_, e2) = Utils.timeTakenMs {
+            filesToDelete.foreach { f =>
+              fm.delete(f.path)
+              val file = new File(rocksDbPath, f.version.toString.toUpperCase(Locale.ROOT))
+              if (file.exists()) {
+                file.delete()
+              }
+            }
+          }
+          logDebug(s"deleting files took $e2 ms.")
+          logInfo(
+            s"Deleted files older than ${earliestFileToRetain.version} for $this: " +
+              filesToDelete.mkString(", "))
+        }
+      }
+    } catch {
+      case NonFatal(e) =>
+        logWarning(s"Error cleaning up files for $this", e)
+    }
+  }
+
+  private def cleanRocksDBBackupInstances(files: Seq[WALUtils.StoreFile]): Unit = {
+    try {
+      if (files.nonEmpty) {
+        val earliestVersionToRetain = files.last.version - storeConf.minVersionsToRetain
+        if (earliestVersionToRetain > 0) {
+          for (v <- (earliestVersionToRetain - 1) to 1) {
+            // Destroy the backup path
+            logDebug((s"Destroying backup version = $v"))
+            RocksDbInstance.destroyDB(getBackupPath(v))
+          }
+        }
+      }
+    } catch {
+      case NonFatal(e) => logWarning(s"Error cleaning up files for $this", e)
+    }
+  }
+
+  // Used only for unit tests
+  private[sql] def latestIterator(): Iterator[UnsafeRowPair] = synchronized {
+    val versionsInFiles = fetchFiles(fm, baseDir).map(_.version).toSet
+    var itr = Iterator.empty
+    if (versionsInFiles.nonEmpty) {
+      val maxVersion = versionsInFiles.max
+      if (maxVersion == 0) {
+        return Iterator.empty
+      }
+      // FIXME assuming maxVersion exists in rocksDB
+      val path = getBackupPath(maxVersion)
+      val r: RocksDbInstance = new RocksDbInstance(keySchema, valueSchema, maxVersion.toString)
+      try {
+        r.open(path, rocksDbConf, readOnly = true)
+        return r.iterator(false)
+      } catch {
+        case e: Exception =>
+        // do nothing
+      }
+    }
+    Iterator.empty
+  }
+
+  // making it public for unit tests
+  lazy val rocksDbPath: String = {
+    val checkpointRootLocationPath = new Path(stateStoreId.checkpointRootLocation)
+    val basePath = new Path(
+      localDirectory,
+      new Path(
+        "db",
+        checkpointRootLocationPath.getName + "_" + checkpointRootLocationPath.hashCode()))
+
+    val dir = basePath.toString + Path.SEPARATOR +
+      stateStoreId_.operatorId + Path.SEPARATOR +
+      stateStoreId_.partitionId
+
+    val f: File = new File(dir)
+
+    if (!f.exists()) {
+      logInfo(s"creating rocksDb directory at : $dir")
+      f.mkdirs()
+    }
+    dir
+  }
+
+  private def getBackupPath(version: Long): String = {
```

Comment:
implementation is same as `getTempPath`. Please deduplicate either one.

Reply:
will do